web/src/app/notificationUtils.js

```diff
@@ -0,0 +1,70 @@
+// This is a separate file since the other utils import `config.js`, which depends on `window`
+// and cannot be used in the service worker
+
+import emojisMapped from "./emojisMapped";
+
+const toEmojis = (tags) => {
+  if (!tags) return [];
+  return tags.filter((tag) => tag in emojisMapped).map((tag) => emojisMapped[tag]);
+};
+
+export const formatTitle = (m) => {
+  const emojiList = toEmojis(m.tags);
+  if (emojiList.length > 0) {
+    return `${emojiList.join(" ")} ${m.title}`;
+  }
+  return m.title;
+};
+
+const formatTitleWithDefault = (m, fallback) => {
+  if (m.title) {
+    return formatTitle(m);
+  }
+  return fallback;
+};
+
+export const formatMessage = (m) => {
+  if (m.title) {
+    return m.message;
+  }
+  const emojiList = toEmojis(m.tags);
+  if (emojiList.length > 0) {
+    return `${emojiList.join(" ")} ${m.message}`;
+  }
+  return m.message;
+};
+
+const isImage = (filenameOrUrl) => filenameOrUrl?.match(/\.(png|jpe?g|gif|webp)$/i) ?? false;
+
+export const icon = "/static/images/ntfy.png";
+export const badge = "/static/images/mask-icon.svg";
+
+export const toNotificationParams = ({ subscriptionId, message, defaultTitle, topicRoute }) => {
+  const image = isImage(message.attachment?.name) ? message.attachment.url : undefined;
```

Comment:
do we maybe want to do `attachment?.type?.startsWith('image/') || attachment?.filename?.match(...)`?